pkg/controller/certificates/sync.go

```diff
@@ -324,3 +338,35 @@ func (c *Controller) updateCertificateStatus(old, new *v1alpha1.Certificate) (*v
 	// for CRDs (https://github.com/kubernetes/kubernetes/issues/38113)
 	return c.CMClient.CertmanagerV1alpha1().Certificates(new.Namespace).Update(new)
 }
+
+func (c *Controller) calculateTimeBeforeExpiry(cert *x509.Certificate, crt *v1alpha1.Certificate) time.Duration {
+	// validate if the certificate received was with the issuer configured
+	// duration. If not we generate an event to warn the user of that fact.
+	certDuration := cert.NotAfter.Sub(cert.NotBefore)
+	if certDuration < crt.Spec.Duration.Duration {
+		s := fmt.Sprintf(messageCertificateDuration, certDuration, crt.Spec.Duration.Duration)
+		glog.Info(s)
+		// TODO Use the message as the reason in a 'renewal status' condition
+	}
+	// renew is the duration before the certificate expiration that cert-manager
+	// will start to try renewing the certificate.
+	renew := v1alpha1.DefaultRenewBefore
+	if crt.Spec.RenewBefore.Duration != 0 {
+		renew = crt.Spec.RenewBefore.Duration
+	}
+	// Verify that the renewBefore duration is inside the certificate validity duration.
+	// If not we notify with an event that we will renew the certificate
+	// before (certificate duration / 3) of its expiration duration.
+	if renew > certDuration {
+		glog.Info(messageScheduleModified)
+		// TODO Use the message as the reason in a 'renewal status' condition
+		// We will renew 1/3 before the expiration date.
+		renew = certDuration / 3
+	}
+	// calculate the amount of time until expiry
+	durationUntilExpiry := cert.NotAfter.Sub(now())
+	// calculate how long until we should start attempting to renew the
+	// certificate
+	renewIn := durationUntilExpiry - renew
+	return renewIn
+}
```

Comment:
Sorry to be a pain, but can we get a unit test for the behaviour of this function? It'll not only make it clearer how cert-manager behaves here, but it'll also make sure that the behaviour doesn't change in future 😄

Reply:
OK there was supposed to be a unit test for this, I think it got lost during some of the refactoring (either moving from issuer to cert or removing events I forget which)